### PR TITLE
Higher Weight Decay: 5e-4 for stronger L2 regularization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The baseline uses `--weight_decay 5e-5`. This is relatively light regularization. With the Lion optimizer (which uses decoupled weight decay), increasing weight decay to **5e-4 (10x)** provides significantly stronger L2 regularization on all model parameters.

**Why this might help:**
1. **Prevents weight magnitude growth.** Over 150 epochs, weights can grow to large magnitudes. Large weights → sharp loss landscape → poor OOD generalization.
2. **Implicit flat minima.** Higher weight decay biases the optimizer toward solutions with smaller parameter norms, which tend to be in flatter regions of the loss landscape (Neyshabur et al., 2017).
3. **Complements cosine annealing.** As LR decays toward zero, the weight decay continues pushing weights toward zero. With higher weight decay, this effect is stronger in the late training phase, providing additional regularization at convergence.
4. **Lion-specific benefit.** Lion uses sign-based updates with decoupled weight decay. Unlike Adam, Lion's weight decay acts independently of the gradient magnitude. A higher rate more aggressively constrains the model.

**Confidence:** Medium. Weight decay is the most fundamental regularizer but 10x increase is aggressive. The risk: if the model needs large weights to represent the physics (e.g., attention scaling, pressure magnitude), too much weight decay will underfit.

## Instructions

**No code changes needed.** Simply change `--weight_decay 5e-5` to `--weight_decay 5e-4` in the run command.

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent nezuko --wandb_name "nezuko/wd-5e4-s42" \
  --wandb_group "round19/higher-weight-decay" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-4 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73 — identical but --seed 73 --wandb_name "nezuko/wd-5e4-s73"
```

Only change: `--weight_decay 5e-4` (was `5e-5`)

### Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| p_tan  | **28.341** | < 28.34 |
| p_re   | **6.300**  | < 6.30  |

W&B: `hgml7i2r` (s42), `qic03vrg` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent nezuko --wandb_name "nezuko/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Higher Weight Decay: 5e-4 (10x baseline)

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed Avg | Delta |
|--------|----------|---------|---------|------------|-------|
| p_in   | **11.979** | 12.0 | 12.8 | 12.40 | +0.42 ❌ |
| p_oodc | **7.643**  | 7.9  | 7.3  | 7.60  | -0.04 ✅ |
| p_tan  | **28.341** | 28.1 | 27.9 | 28.00 | -0.34 ✅ |
| p_re   | **6.300**  | 6.5  | 6.4  | 6.45  | +0.15 ❌ |

**W&B runs:**
- Seed 42: [`w15nt6uv`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/w15nt6uv)
- Seed 73: [`jjayhwof`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/jjayhwof)

**Commands used:**
```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --wandb_name "nezuko/wd-5e4-s42" \
  --wandb_group "round19/higher-weight-decay" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-4 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73 — identical but --seed 73 --wandb_name "nezuko/wd-5e4-s73"
```

Both best models at epoch 148 (hit the epoch limit, stopped by timeout).

### What happened

**Mixed results — overall negative.** The 10x weight decay increase (5e-5 → 5e-4) shows a clear trade-off:

- **OOD conditions (p_oodc) improved slightly** (-0.04), and **tandem transfer (p_tan) improved** (-0.34). This is consistent with the hypothesis that stronger weight decay encourages flatter minima with better OOD generalization.
- **In-distribution (p_in) degraded significantly** (+0.42 avg), and **Reynolds OOD (p_re) also worsened** (+0.15). The in-distribution degradation on seed 73 (12.8 vs 11.979 baseline) is particularly notable.
- **High seed variance**: p_oodc ranges from 7.3 to 7.9 across seeds, and p_in from 12.0 to 12.8. This suggests the stronger weight decay may be making optimization less stable.

The hypothesis that higher weight decay prevents weight magnitude growth and finds flatter minima is partially confirmed for OOD generalization, but the cost to in-distribution accuracy is too high. The 10x jump was likely too aggressive — the model is being over-regularized, constraining its capacity to represent the physics accurately on the training distribution.

### Suggested follow-ups

1. **Moderate weight decay (1e-4 or 2e-4)** — A 2-4x increase instead of 10x might capture the OOD benefits without the in-distribution penalty.
2. **Scheduled weight decay** — Start with 5e-5 and ramp to 5e-4 in the later training phase (when cosine LR is already low), so the model first learns the physics then gets regularized.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2246 because ADVISOR: Closing — mixed result, net negative. p_tan improved (-1.2%) and p_oodc marginal (-0.6%), but p_in regressed significantly (+3.5%) and p_re worsened (+2.4%). High seed variance (p_in: 12.0 vs 12.8, p_oodc: 7.3 vs 7.9) makes the directional signal unreliable. 10x weight decay was too aggressive — over-regularization constrained capacity to fit in-distribution physics. Moderate increase (1e-4, 2x) is noted as a potential direction but the variance suggests this is unlikely to be a productive lever. Current 5e-5 appears well-calibrated..
